### PR TITLE
Add ML-EM and OS-EM methods and examples

### PR DESCRIPTION
**Pull request by ZongyuLi-umich**
need to first fix the project index bug before running the osem example.

### CODE REVIEW

File: docs/lit/examples/6-dl.jl [hunk 1]
```diff
@@ -302,7 +303,7 @@ jim(
     jim(mid3(xhat1), "EM recon, NRMSE = $(nrmse(xhat1))%"; clim),
     jim(mid3(xiter1), "Iter 1, NRMSE = $(nrmse(xiter1))%"; clim),
     jim(mid3(xiter2), "Iter 2, NRMSE = $(nrmse(xiter2))%"; clim),
-)
+    )
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:306]:
```suggestion
)
```
See the section on Function calls here:
https://github.com/invenia/BlueStyle#whitespace

File: docs/lit/examples/7-osem.jl [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+end
+jim(mid3(xtrue), "Middle slices of xtrue")
+
+
+# ### PSF
+
+# Create a synthetic depth-dependent PSF for a single view
+px = 11
+psf1 = psf_gauss( ; ny, px)
+jim(psf1, "PSF for each of $ny planes")
```
Comment (JeffFessler) [inline docs/lit/examples/7-osem.jl:52]:
```suggestion
jim(psf1, "PSF for each of $ny planes"; ratio=1)
```
This is a new option that looks better here :)

File: docs/lit/examples/7-osem.jl [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		asum[nb] = Ab[nb]' * ones(eltype(ynoisy), nx, nz, nview÷nblocks)
+	end
+	time0 = time()
+    for iter = 1:niter
+		for nb = 1:nblocks
+			@show iter, nb, extrema(x), time() - time0
+			ybar = Ab[nb] * x .+ (@view background[:,:,nb:nblocks:nview]) # forward model
+			x .*= (Ab[nb]' * ((@view ynoisy[:,:,nb:nblocks:nview]) ./ ybar)) ./ asum[nb] # multiplicative update
+		end
+    end
```
Comment (JeffFessler) [inline docs/lit/examples/7-osem.jl:127]: something awry with indenting.  use 4 spaces, not tab?

File: src/backproject.jl [hunk 4]
```diff
@@ -139,17 +139,17 @@ function backproject!(
     image .= zero(plan.T) # must be initialized as zero
     if plan.mode === :fast
         [plan.add_img[i] .= zero(plan.T) for i = 1:plan.nthread]
-        Threads.@threads for i in index
+        Threads.@threads for (i, viewidx) in collect(enumerate(index))
```
Comment (JeffFessler) [inline src/backproject.jl:142]:
```suggestion
        Threads.@threads for (i, viewidx) in enumerate(index)
```
collect is not needed

Comment (ZongyuLi-umich) [reply src/backproject.jl:142]: `@threads` does not work with generic iterative variables like`enumerate`. See [this](https://discourse.julialang.org/t/enumerate-not-suported-for-threads/37670).

Comment (JeffFessler) [reply src/backproject.jl:142]: Oh!  OK.  I will retract the suggestion.

File: docs/lit/examples/6-dl.jl [hunk 5]
```diff
@@ -40,6 +40,7 @@ using Distributions: Poisson
 using BSON: @load, @save
 import BSON # load
 using InteractiveUtils: versioninfo
+using Downloads: download
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:43]: good suggestion - i didn't know about this.
but then you must use `Downloads.download` below

File: Project.toml [hunk 6]
```diff
@@ -14,7 +14,7 @@ OffsetArrays = "6fe1bfb0-de20-5000-8ca7-80f57d26f881"
 [compat]
 AbstractFFTs = "1.0"
 FFTW = "1.4"
-ImageFiltering = "0.6"
+ImageFiltering = "0.7"
```
Comment (JeffFessler) [inline Project.toml:17]:
```suggestion
ImageFiltering = "0.6, 0.7"
```
let's see if this fixes the test failures.

File: src/ml-os-em.jl [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+- `ynoisy`: (Noisy) measurements
+- `background`: Background effects, e.g., scatters
+- `A`: System matrix
+- `niter`: Number of iterations
+"""
+mlem(x0::AbstractArray,
+	 ynoisy::AbstractArray,
+	 background::AbstractArray,
+	 A::Union{AbstractArray, LinearMapAO};
+	 kwargs...) = mlem!(similar(x0), x0, ynoisy, background, A; kwargs...)
```
Comment (JeffFessler) [inline src/ml-os-em.jl:57]:
```suggestion
function mlem(
    x0::AbstractArray,
    ynoisy::AbstractArray,
    background::AbstractArray,
    A::Union{AbstractArray, LinearMapAO};
    kwargs...,
 )
     return mlem!(similar(x0), x0, ynoisy, background, A; kwargs...)
 end
```

File: src/ml-os-em.jl [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        mul!(back, A', yratio) # back = A' * (ynoisy / ybar)
+        @. out *= back / asum # multiplicative update
+    end
+    return out
+end
+
+
+"""
+    mlem(x0, ynoisy, background, A; niter = 20)
+ML-EM algorithm for SPECT reconstruction.
```
Comment (JeffFessler) [inline src/ml-os-em.jl:46]:
```suggestion
ML-EM algorithm for emission tomography image reconstruction.
```

File: src/ml-os-em.jl [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+- `background`: Background effects, e.g., scatters
+- `A`: System matrix
+- `niter`: Number of iterations
+"""
+function mlem!(out::AbstractArray,
+			   x0::AbstractArray,
+               ynoisy::AbstractArray,
+               background::AbstractArray,
+               A::Union{AbstractArray, LinearMapAO};
+               niter::Int = 20)
```
Comment (JeffFessler) [inline src/ml-os-em.jl:23]:
```suggestion
function mlem!(
    out::AbstractArray,
    x0::AbstractArray,
    ynoisy::AbstractArray,
    background::AbstractArray,
    A::Union{AbstractArray, LinearMapAO};
    niter::Int = 20,
 )
```

File: src/ml-os-em.jl [hunk 10]
```diff
@@ -0,0 +1,140 @@
+# mlem.jl
+# ML-EM algortihm for SPECT reconstruction
+export mlem, mlem!, osem, osem!
+export Ablock
+using LinearMapsAA: LinearMapAO, LinearMapAA
+
+
+"""
+    mlem!(out, x0, ynoisy, background, A; niter = 20)
+Inplace version of ML-EM algorithm for SPECT reconstruction.
```
Comment (JeffFessler) [inline src/ml-os-em.jl:10]:
```suggestion
Inplace version of ML-EM algorithm for emission tomography image reconstruction.
```

File: src/ml-os-em.jl [hunk 11]
```diff
@@ -0,0 +1,140 @@
+# mlem.jl
```
Comment (JeffFessler) [inline src/ml-os-em.jl:1]:
```suggestion
# ml-os-em.jl
```

File: src/ml-os-em.jl [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+# ML-EM algortihm for SPECT reconstruction
```
Comment (JeffFessler) [inline src/ml-os-em.jl:2]:
```suggestion
# ML-EM algorithm for emission tomography image reconstruction
```

File: src/ml-os-em.jl [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+- `niter`: Number of iterations
+"""
+function mlem!(out::AbstractArray,
+			   x0::AbstractArray,
+               ynoisy::AbstractArray,
+               background::AbstractArray,
+               A::Union{AbstractArray, LinearMapAO};
+               niter::Int = 20)
+    all(>(0), background) || throw("need background > 0")
+	size(out) != size(x0) && throw(DimensionMismatch("size out and x0 not match"))
```
Comment (JeffFessler) [inline src/ml-os-em.jl:25]:
```suggestion
    size(out) == size(x0) || throw(DimensionMismatch("size out and x0 not match"))
```

File: src/ml-os-em.jl [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    asum = A' * ones(eltype(ynoisy), size(ynoisy)) # this allocates
+    asum[(asum .== 0)] .= Inf
+    ybar = similar(ynoisy)
+    yratio = similar(ynoisy)
+    back = similar(x0)
+	copyto!(out, x0)
```
Comment (JeffFessler) [inline src/ml-os-em.jl:31]:
```suggestion
    copyto!(out, x0)
```

File: src/ml-os-em.jl [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    time0 = time()
+    for iter = 1:niter
+        @show iter, extrema(out), time() - time0
```
Comment (JeffFessler) [inline src/ml-os-em.jl:34]:
```suggestion
        chat && (@show iter, extrema(out), time() - time0)
```

File: src/ml-os-em.jl [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+"""
+    Ablock(plan, nblocks)
+Generate a vector of linear maps for OSEM.
+-`plan`: SPECTrecon plan
+-`nblocks`: Number of blocks in OSEM
+"""
+function Ablock(plan::SPECTplan, nblocks::Int)
+    nx, ny, nz = size(plan.mumap)
+    nview = plan.nview
+    @assert rem(nview, nblocks) == 0 || throw("nview must be divisible by nblocks!")
```
Comment (JeffFessler) [inline src/ml-os-em.jl:69]:
```suggestion
    rem(nview, nblocks) == 0 || throw("nview must be divisible by nblocks!")
```
we don't use both assert and throw together

File: src/ml-os-em.jl [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+- `background`: Background effects, e.g., scatters
+- `Ab`: Vector of system matrix
+- `niter`: Number of iterations
+"""
+function osem!(out::AbstractArray,
+			   x0::AbstractArray,
+               ynoisy::AbstractArray,
+               background::AbstractArray,
+               Ab::Union{Vector{AbstractArray}, Vector{LinearMapAO}};
+               niter::Int = 16)
```
Comment (JeffFessler) [inline src/ml-os-em.jl:98]:
```suggestion
function osem!(
    out::AbstractArray,
    x0::AbstractArray,
    ynoisy::AbstractArray,
    background::AbstractArray,
    Ab::Union{Vector{AbstractArray}, Vector{LinearMapAO}};
    niter::Int = 16,
    chat::Bool = false,
)
```

File: src/ml-os-em.jl [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    all(>(0), background) || throw("need background > 0")
+	size(out) != size(x0) && throw(DimensionMismatch("size out and x0 not match"))
+	nx, nz, nview = size(ynoisy)
+	nblocks = length(Ab)
+	asum = Vector{Array{eltype(ynoisy), 3}}(undef, nblocks)
+	for nb = 1:nblocks
+	    asum[nb] = Ab[nb]' * ones(eltype(ynoisy), nx, nz, nview÷nblocks)
+        (asum[nb])[(asum[nb] .== 0)] .= Inf # avoid divide by zero
+	end
```
Comment (JeffFessler) [inline src/ml-os-em.jl:107]:
```suggestion
    size(out) != size(x0) && throw(DimensionMismatch("size out and x0 not match"))
    nx, nz, nview = size(ynoisy)
    nblocks = length(Ab)
    asum = Vector{Array{eltype(ynoisy), 3}}(undef, nblocks)
    for nb = 1:nblocks
        asum[nb] = Ab[nb]' * ones(eltype(ynoisy), nx, nz, nview÷nblocks)
        (asum[nb])[asum[nb] .== 0] .= Inf # avoid divide by zero
    end
```

File: src/ml-os-em.jl [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    ybar = Array{eltype(ynoisy)}(undef, nx, nz, nview÷nblocks)
+    yratio = similar(ybar)
+    back = similar(x0)
+	copyto!(out, x0)
+	time0 = time()
```
Comment (JeffFessler) [inline src/ml-os-em.jl:112]:
```suggestion
    copyto!(out, x0)
    time0 = time()
```

File: src/ml-os-em.jl [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+    for iter = 1:niter
+        for nb = 1:nblocks
+	        @show iter, nb, extrema(out), time() - time0
```
Comment (JeffFessler) [inline src/ml-os-em.jl:115]:
```suggestion
	        chat && (@show iter, nb, extrema(out), time() - time0)
```

Comment (JeffFessler) [reply src/ml-os-em.jl:115]: please fix tab with 4space around here...

File: src/ml-os-em.jl [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+- `ynoisy`: (Noisy) measurements
+- `background`: Background effects, e.g., scatters
+- `Ab`: Vector of system matrix
+- `niter`: Number of iterations
+"""
+osem(x0::AbstractArray,
+	 ynoisy::AbstractArray,
+	 background::AbstractArray,
+	 Ab::Union{Vector{AbstractArray}, Vector{LinearMapAO}};
+	 kwargs...) = osem!(similar(x0), x0, ynoisy, background, Ab; kwargs...)
```
Comment (JeffFessler) [inline src/ml-os-em.jl:140]:
```suggestion
osem(x0::AbstractArray, args..., kwargs...) = osem!(similar(x0), x0, args...; kwargs...)
```

File: src/ml-os-em.jl [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+"""
+    osem(x0, ynoisy, background, Ab; niter = 20)
+OS-EM algorithm for SPECT reconstruction.
+- `x0`: Initial guess
+- `ynoisy`: (Noisy) measurements
+- `background`: Background effects, e.g., scatters
+- `Ab`: Vector of system matrix
+- `niter`: Number of iterations
+"""
+osem(x0::AbstractArray, args..., kwargs...) = osem!(similar(x0), x0, args...; kwargs...)
```
Comment (JeffFessler) [inline src/ml-os-em.jl:141]:
```suggestion
osem(x0::AbstractArray, args...; kwargs...) = osem!(similar(x0), x0, args...; kwargs...)
```

File: src/ml-os-em.jl [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+"""
+    mlem(x0, ynoisy, background, A; niter = 20)
+ML-EM algorithm for emission tomography image reconstruction.
+- `x0`: Initial guess
+- `ynoisy`: (Noisy) measurements
+- `background`: Background effects, e.g., scatters
+- `A`: System matrix
+- `niter`: Number of iterations
+"""
+mlem(x0::AbstractArray, args...; kwargs...) = mlem!(similar(x0), args...; kwargs...)
```
Comment (JeffFessler) [inline src/ml-os-em.jl:59]:
```suggestion
mlem(x0::AbstractArray, args...; kwargs...) = mlem!(similar(x0), x0, args...; kwargs...)
```

**Comment by JeffFessler** [comment]
I broke it so I should fix it!

File: docs/lit/examples/4-mlem.jl [hunk 24]
```diff
@@ -87,62 +87,25 @@ if !@isdefined(ynoisy) # generate (scaled) Poisson data
     scale = target_mean / average(ytrue)
     scatter_fraction = 0.1 # 10% uniform scatter for illustration
     scatter_mean = scatter_fraction * average(ytrue) # uniform for simplicity
+    background = scatter_mean * ones(T,nx,nz,nview)
     ynoisy = rand.(Poisson.(scale * (ytrue .+ scatter_mean))) / scale
```
Comment (JeffFessler) [inline docs/lit/examples/4-mlem.jl:91]:
```suggestion
    ynoisy = rand.(Poisson.(scale * (ytrue + background))) / scale
```

File: docs/lit/examples/6-dl.jl [hunk 25]
```diff
@@ -118,40 +119,20 @@ if !@isdefined(ynoisy) # generate (scaled) Poisson data
     scale = target_mean / average(ytrue)
     scatter_fraction = 0.1 # 10% uniform scatter for illustration
     scatter_mean = scatter_fraction * average(ytrue) # uniform for simplicity
+    background = scatter_mean * ones(T,nx,nz,nview)
     ynoisy = rand.(Poisson.(scale * (ytrue .+ scatter_mean))) / scale
```
Comment (JeffFessler) [inline docs/lit/examples/6-dl.jl:123]:
```suggestion
    ynoisy = rand.(Poisson.(scale * (ytrue + background))) / scale
```

File: docs/lit/examples/7-osem.jl [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+# Noisy data
+
+if !@isdefined(ynoisy) # generate (scaled) Poisson data
+    ytrue = A * xtrue
+    target_mean = 20 # aim for mean of 20 counts per ray
+    scale = target_mean / average(ytrue)
+    scatter_fraction = 0.1 # 10% uniform scatter for illustration
+    scatter_mean = scatter_fraction * average(ytrue) # uniform for simplicity
+	background = scatter_mean * ones(T,nx,nz,nview)
+    ynoisy = rand.(Poisson.(scale * (ytrue .+ scatter_mean))) / scale
```
Comment (JeffFessler) [inline docs/lit/examples/7-osem.jl:86]:
```suggestion
    background = scatter_mean * ones(T,nx,nz,nview)
    ynoisy = rand.(Poisson.(scale * (ytrue + background))) / scale
```

File: test/ml-os-em.jl [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+    odim = (nx,nz,nview)
+    A = LinearMapAA(forw!, back!, (prod(odim),prod(idim)); T, odim, idim)
+    ytrue = A * xtrue
+    target_mean = 20 # aim for mean of 20 counts per ray
+    average(x) = sum(x) / length(x)
+    scale = target_mean / average(ytrue)
+    scatter_fraction = 0.1 # 10% uniform scatter for illustration
+    scatter_mean = scatter_fraction * average(ytrue) # uniform for simplicity
+    ynoisy = rand.(Poisson.(scale * (ytrue .+ scatter_mean))) / scale
+    background = scatter_mean * ones(T, nx, nz, nview) # uniform scatter estimation
```
Comment (JeffFessler) [inline test/ml-os-em.jl:39]:
```suggestion
    background = scatter_mean * ones(T, nx, nz, nview) # uniform scatter
    ynoisy = rand.(Poisson.(scale * (ytrue + background))) / scale

```

**Pull request merged**